Foxtrot/Driver/Rewriting/EmitAsyncClosure.cs

```diff
@@ -59,6 +58,121 @@ namespace Microsoft.Contracts.Foxtrot
+        /// IL_0016: box !!0 // <-- here is our problem!
+        /// ]]>
+        /// </code>
+        ///
+        /// This means that method <code>CheckPost</code> would contains a reference to generic method argument of the
+        /// original method.
+        /// 
+        /// The goal of this class is to store a mapping between enclosing generic types and closure generic types.
+        /// </remarks>
+        private class GenericTypeMapper
+        {
+            class TypeNodePair
+            {
+                public TypeNodePair(TypeNode enclosingGenericType, TypeNode closureGenericType)
+                {
+                    EnclosingGenericType = enclosingGenericType;
+                    ClosureGenericType = closureGenericType;
+                }
+
+                public TypeNode EnclosingGenericType { get; private set; }
+                public TypeNode ClosureGenericType { get; private set; }
+            }
+
+            // Mapping between enclosing generic type and closure generic type.
+            // This is a simple list not a dictionary, because number of generic arguments is very small.
+            // So linear complexity will not harm performance.
+            private List<TypeNodePair> typeParametersMapping = new List<TypeNodePair>();
+
+            public bool IsEmpty { get { return typeParametersMapping.Count == 0; } }
+
+            public void AddMapping(TypeNode enclosingGenericType, TypeNode closureGenericType)
+            {
+                typeParametersMapping.Add(new TypeNodePair(enclosingGenericType, closureGenericType));
+            }
+
+            /// <summary>
+            /// Returns associated generic type of the closure class by enclosing generic type (for instance, by
+            /// generic type of the enclosing generic method that uses current closure).
+            /// </summary>
+            /// <remarks>
+            /// Function returns the same argument if the matching argument does not exists.
+            /// </remarks>
+            public TypeNode GetClosureTypeParameterByEnclosingTypeParameter(TypeNode enclosingType)
+            {
+                if (enclosingType == null)
+                {
+                    return null;
+                }
+
+                var candidate = typeParametersMapping.FirstOrDefault(t => t.EnclosingGenericType == enclosingType);
+                return candidate != null ? candidate.ClosureGenericType : enclosingType;
+            }
+            
+            /// <summary>
+            /// Returns associated generic type of the closure class by enclosing generic type (for instance, by
+            /// generic type of the enclosing generic method that uses current closure).
+            /// </summary>
+            /// <remarks>
+            /// Function returns the same argument if the matching argument does not exists.
+            /// </remarks>
+            public TypeNode GetEnclosingTypeParamterByGenericTypeParameter(TypeNode closureType)
```

Comment:
I am wondering if the intention was to name the method that way: GetEnclosingTypeParamterBy**Closure**TypeParameter